### PR TITLE
[FLINK-28853][connector-base] Add FLIP-217 support for watermark alignment of source splits

### PR DESCRIPTION
**Pull request by smattheis**
## What is the purpose of the change

* The change implements FLIP-217 (https://cwiki.apache.org/confluence/display/FLINK/FLIP-217+Support+watermark+alignment+of+source+splits) to support watermark alignment of source splits.

## Brief change log

* Revised `SplitFetcher` threading model and add support to pause/resume `SplitFetcher`
* Adds support for watermark alignment of individual source splits controlled by `SourceOperator`
* Adds support for split watermark alignment for Kafka and Pulsar sources
* Adds configuration parameter to allow unaligned splits as migration plan to support legacy sources that lack support for split alignment

## Verifying this change

This change added and extended tests and can be verified as follows:

* `SplitFetcherTest` extended to test pause/resume of`SplitFetcher`
* `SourceOperatorSplitWatermarkAlignmentTest` added to test split alignment based on split reader watermarks
* `SplitFetcherPauseResumeSplitReaderITCase` added to test pause/resume of individual split readers

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes** (according to FLIP-217)
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **docs** (configuration parameter) and **JavaDocs** (changes of public APIs)

### CODE REVIEW

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherPauseResumeSplitReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                    split.addRecord(i * numSplits + splitId);
+                }
+                splits.add(split);
+            }
+            return splits;
+        }
+
+        public void addPrefilledSplitsSingleReader(int numSplits, int numRecords) {
+            sourceReader.addSplits(createPrefilledSplits(numSplits, numRecords));
+            sourceReader.notifyNoMoreSplits();
```
Comment (mas-chen) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherPauseResumeSplitReaderTest.java:247]: Is this a required assumption? We should also exercise the case where splits are dynamically discovered (e.g. Kafka source topic partition discovery).

**Review by mas-chen — COMMENTED** [review]
I did a quick first pass and still need to look deeper into the SplitFetcher changes. Lots of code changes! Do we have performance tests with analysis on the split fetcher changes, even without split alignment enabled?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java [hunk 2]
```diff
@@ -255,6 +256,23 @@ public void testClose() {
...
+                createFetcherWithSplit(
+                        "test-split",
+                        queue,
+                        new TestingSplitReader<>(finishedSplitFetch("test-split")));
+
+        fetcher.pause();
+
+        new Thread(fetcher::shutdown).start();
+
+        fetcher.runOnce();
```
Comment (mas-chen) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java:273]: Can we assert that this is false?

**Comment by mxm** [comment]
Any thoughts on how and when to proceed with this PR? We'd like to help to bring this over the finish line.

**Comment by pnowojski** [comment]
> I wonder about the way the SourceOperator calls through the stack via pauseOrResumeSplits while also being called back with watermarks. My intuition would have been to do the pausing directly at the splits based on the current max watermark lag, e.g. by putting splits into a suspended mode where no more output is accepted from a split. I guess this could be tricky because the implementation of the source doesn't necessarily know it's paused and could timeout or otherwise misbehave. So it looks like the current implementation is simpler.

Currently Splits are just pojos, data carriers, without much logic. If I understand you correctly, what you are suggesting would require us to push some availability logic down to the splits and it would still require handling of this availability in the `SourceReader` - basically the same logic that we have in the current proposal. The difference would be only that `SourceReader` would be informed about (un)availability change of a split, from the split, instead of from the `pauseOrResume` call.

Given that most of the users are expected to just use `SourceReaderBase`, the current proposal gives them watermark alignment for free. While adding availability logic to the splits would increase exposure of this feature in the public API, without much benefit?

> Any thoughts on how and when to proceed with this PR? We'd like to help to bring this over the finish line.

Indeed there is an issue now, that I think @smattheis doesn't work on Flink anymore, so he can not finish this PR. I will go through the PR myself and do a quick pass/respond to the inlined questions from you @mxm  and @mas-chen , but it would be great if someone could take over applying the actual changes after the review. Would you be up to do that @mxm ?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 3]
```diff
@@ -559,14 +577,67 @@ private void createOutputForSplits(List<SplitT> newSplits) {
...
+        lastEmittedWatermark = watermark;
         checkWatermarkAlignment();
     }
 
+    @Override
+    public void updateCurrentSplitWatermark(String splitId, long watermark) {
+        splitCurrentWatermarks.put(splitId, watermark);
+        if (currentMaxDesiredWatermark < watermark && !currentlyPausedSplits.contains(splitId)) {
+            pauseOrResumeSplits(Collections.singletonList(splitId), Collections.emptyList());
+            currentlyPausedSplits.add(splitId);
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:595]: ? This line that you are pointing to 
```
plitsToPause.removeAll(currentlyPausedSplits);
```
doesn't modify `currentlyPausedSplits`. It only gathers the information what splits have to be paused/resumed.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 4]
```diff
@@ -559,14 +577,67 @@ private void createOutputForSplits(List<SplitT> newSplits) {
...
+    /**
+     * Finds the splits that are beyond the current max watermark and pauses them. At the same time,
+     * splits that have been paused and where the global watermark caught up are resumed.
+     *
+     * <p>Note: This takes effect only if there are multiple splits, otherwise it does nothing.
+     */
+    private void checkSplitWatermarkAlignment() {
+        if (numSplits <= 1) {
+            return; // If there is only a single split, we do not pause the split but the source.
+        }
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:608]: Would you be willing to simplify this? If so, I would suggest to do it as a separate commit (either in the same PR or a new PR).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 5]
```diff
@@ -559,14 +577,67 @@ private void createOutputForSplits(List<SplitT> newSplits) {
...
+                    } else if (currentlyPausedSplits.contains(splitId)) {
+                        splitsToResume.add(splitId);
+                    }
+                });
+        splitsToPause.removeAll(currentlyPausedSplits);
+        if (!splitsToPause.isEmpty() || !splitsToResume.isEmpty()) {
+            pauseOrResumeSplits(splitsToPause, splitsToResume);
+            currentlyPausedSplits.addAll(splitsToPause);
+            splitsToResume.forEach(currentlyPausedSplits::remove);
+        }
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:624]: What do you mean "signal to user"?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java [hunk 6]
```diff
@@ -227,6 +227,20 @@ public void close() throws Exception {
...
+            Collection<KafkaPartitionSplit> splitsToPause,
+            Collection<KafkaPartitionSplit> splitsToResume) {
+        consumer.resume(
+                splitsToResume.stream()
+                        .map(KafkaPartitionSplit::getTopicPartition)
+                        .collect(Collectors.toList()));
+        consumer.pause(
+                splitsToPause.stream()
+                        .map(KafkaPartitionSplit::getTopicPartition)
+                        .collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:241]: I'm not sure how likely is this `resume` to ever block and even if it blocks, I don't think it should cause any problems. Keep in mind that this is not an exact feature. We do not guarantee that no records will be processed . The whole watermark alignment requires sending RPCs between TMs and JM so it has built in delay/lag.

At the same time I don't see how doing this in the other way around would cause much harm. At worst the consumer would have a for a very short period of time no splits to process if everything is paused but not yet resumed.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderBase.java [hunk 7]
```diff
@@ -179,6 +180,21 @@ public void handleSplitsChanges(SplitsChange<PulsarPartitionSplit> splitsChanges
         this.pulsarConsumer = consumer;
     }
 
+    @Override
+    public void pauseOrResumeSplits(
+            Collection<PulsarPartitionSplit> splitsToPause,
+            Collection<PulsarPartitionSplit> splitsToResume) {
+        if (splitsToPause.size() > 1 || splitsToResume.size() > 1) {
+            throw new IllegalStateException("This pulsar split reader only support one split.");
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderBase.java:188]: Good question...

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java [hunk 8]
```diff
@@ -282,4 +282,21 @@ public enum VertexDescriptionMode {
                     .withDescription(
                             "Whether name of vertex includes topological index or not. "
                                     + "When it is true, the name will have a prefix of index of the vertex, like '[vertex-0]Source: source'. It is false by default");
+
+    @PublicEvolving
+    public static final ConfigOption<Boolean> ALLOW_UNALIGNED_SOURCE_SPLITS =
+            key("pipeline.watermark-alignment.allow-unaligned-source-splits")
+                    .booleanType()
+                    .defaultValue(false)
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java:290]: Please check the dev mailing list discussion. We want to change the the old behaviour from FLINK-1.14/1.15. In a couple of releases (1.18? 1.19?), we will want to disallow partially working watermark alignment when source doesn't support per split alignment. At the same time we want to provide users time to adjust/migrate if they are relaying on this partially working feature for now.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 9]
```diff
@@ -559,14 +577,67 @@ private void createOutputForSplits(List<SplitT> newSplits) {
...
+
+    private void pauseOrResumeSplits(
+            Collection<String> splitsToPause, Collection<String> splitsToResume) {
+        if (!allowUnalignedSourceSplits || sourceReaderSupportsPauseOrResumeSplits) {
+            try {
+                sourceReader.pauseOrResumeSplits(splitsToPause, splitsToResume);
+            } catch (UnsupportedOperationException e) {
+                sourceReaderSupportsPauseOrResumeSplits = false;
+                if (!allowUnalignedSourceSplits) {
+                    throw new UnsupportedOperationException("", e);
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:635]: > In my understanding, readers which do not support pausing splits will fail the job, unless the configuration to allow unaligned splits is set. Hence, I don't understand the sourceReaderSupportsPauseOrResumeSplits flag.

please check my comment above.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:624]: I thought we might expose DataInputStatus of the SourceOperator somewhere. Semantically, one split being paused should yield the same status as multiple splits being all paused.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:595]: We do modify `currentlyPausedSplits` two lines down: https://github.com/apache/flink/pull/20485/files#diff-eba14821fb3e96f6f20e3116ceab893c2f18cff605b177dad485aadc43ac4f56R622

**Review by mxm — COMMENTED** [review]
@pnowojski Thank you for the comments. If nobody of the authors currently has time to work on this, I can take over the PR for now. I'd be glad to get another review from @mas-chen and you once I've addressed the comments.

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java:290]: I've read the mailing list thread. I still think it would be better to not enforce this feature for now. If I understand correctly, watermark alignment without split alignment still works but it is possible for splits to overtake each other which can slow down the processing due to regular watermark alignment. I would take it to the mailing list if we want to revisit this decision.

File: docs/layouts/shortcodes/generated/pipeline_configuration.html [hunk 10]
```diff
@@ -128,5 +128,11 @@
             <td>Boolean</td>
             <td>Whether name of vertex includes topological index or not. When it is true, the name will have a prefix of index of the vertex, like '[vertex-0]Source: source'. It is false by default</td>
         </tr>
+        <tr>
+            <td><h5>pipeline.watermark-alignment.allow-unaligned-source-splits</h5></td>
```
[the remark this answers is not in the document]
Comment (mxm) [reply docs/layouts/shortcodes/generated/pipeline_configuration.html:132]: IMHO something like `source.reader.synchronize.splits` would be more understandable for users.

**Comment by mxm** [comment]
It looks like I will have to open a new PR because @smattheis hasn't checked the option "Allow edits by maintainers". I can't push to the branch.

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java:290]: > watermark alignment without split alignment still works but it is possible for splits to overtake each other which can slow down the processing due to regular watermark alignment

I'm not sure if I understand you here. So let me rephrase the consensus.

On the current master. Watermark alignment works to a small extent if a source operator has more than one split, but it doesn't work fully. In worst case, it won't work at all. Imagine if the slowest split, that is a split that's the most lagging/blocking watermark emission, is processed on the same operator that has the fastest split. In such scenario, watermark alignment without per split alignment, can not block processing of this whole operator (or otherwise it would deadlock the watermarks emission), so the fastest split will continue producing records, for example blowing up time windowed stated downstream. Even in average case, such watermark alignment will work very badly. 

Moreover, using pre-FLIP-217 watermark alignment with multiple splits per operator has been always discouraged. With FLIP-217 the consensus was, that it would be actually hurtful for the users, to have _SILENTLY_ hit such scenarios: user could configure watermark alignment, but actually it wouldn't work. The consensus was that from now on this feature should either work fully, or not at all if users tries to enable it. But we also want to provide gradual migration path for any users that would be affected by this change. Instead of forcing user to both upgrade Flink version AND solve the lack of per-split alignment on his side, he can temporarily set `allow-unaligned-source-splits` to `true`, do the Flink upgrade, and then in the background prepare for next Flink release that will remove this option and set it to `false` in stone.

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java [hunk 11]
```diff
@@ -282,4 +282,21 @@ public enum VertexDescriptionMode {
                     .withDescription(
                             "Whether name of vertex includes topological index or not. "
                                     + "When it is true, the name will have a prefix of index of the vertex, like '[vertex-0]Source: source'. It is false by default");
+
+    @PublicEvolving
+    public static final ConfigOption<Boolean> ALLOW_UNALIGNED_SOURCE_SPLITS =
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java:287]: This should be marked as `@Deprecated("Will be removed in future Flink releases")`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:624]: 👍 So as user you mean the flink-runtime? :) 

I think should be indeed checked and asserted via a test. I would expect this to work - if all splits are paused, eventually (maybe after a one or two extra polls) the source should return no more input available.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:595]: What problem do you see with this? The logic of both of those methods looks ok and consistent to me. Are you worried about threading issues? Both of those methods are called only from the task thread.

**Review by pnowojski — COMMENTED** [review]
> If nobody of the authors currently has time to work on this, I can take over the PR for now.

That would be great :)

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java (hunk 11 above)
Comment (mxm) [reply flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java:287]: Deprecated doesn't allow for a string as argument but I've added an additional comment.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java (hunk 2 above)
Comment (mxm) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherTest.java]: Done.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:594]: Thanks! I was worried about threading issues but I see that it's safe due to the single-threaded execution model.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReader.java [hunk 12]
```diff
@@ -132,4 +139,28 @@ default void handleSourceEvents(SourceEvent sourceEvent) {}
...
+     * releases, it is recommended to implement this method and override the default implementation.
+     *
+     * @param splitsToPause the splits to pause
+     * @param splitsToResume the splits to resume
+     */
+    @PublicEvolving
+    default void pauseOrResumeSplits(
+            Collection<String> splitsToPause, Collection<String> splitsToResume) {
+        throw new UnsupportedOperationException(
+                "This source reader does not support pause or resume splits.");
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReader.java:164]: Shouldn't this be kept in sync with the other similar `throw new UnsupportedOperationException` in `SplitReader`?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitReader.java [hunk 13]
```diff
@@ -68,4 +69,29 @@
...
+     *
+     * @param splitsToPause the splits to pause
+     * @param splitsToResume the splits to resume
+     */
+    default void pauseOrResumeSplits(
+            Collection<SplitT> splitsToPause, Collection<SplitT> splitsToResume) {
+        throw new UnsupportedOperationException(
+                "This split reader does not support pause or resume. (To use watermark alignment "
+                        + "and allow unaligned source splits set configuration parameter "
+                        + "'pipeline.watermark-alignment.allow-unaligned-source-splits' to true.)");
```
Comment (pnowojski) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitReader.java:95]: > This split reader does not support pause or resume. It is highly discourage to use
> watermark alignment and allow unaligned source splits, as this leads to unpredictable
> effectiveness of the watermark alignment. It is recommended to implement pausing splits
> for this source. Temporarily you can allow unaligned source splits by setting 
> configuration parameter `pipeline.watermark-alignment.allow-unaligned-source-splits' to true.
> Beware that this configuration parameter will be dropped in a future Flink release.

?

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-core/src/main/java/org/apache/flink/configuration/PipelineOptions.java:291]: I fully agree. I've left a suggestion in this regard.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitReader.java (hunk 13 above)
Comment (pnowojski) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitReader.java]: Sounds great!

> "watermark alignment if there is more than a single split."

change to:

> "watermark alignment if there is more than a single split per split reader."

?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 14]
```diff
@@ -559,14 +577,67 @@ private void createOutputForSplits(List<SplitT> newSplits) {
...
+    public void updateCurrentEffectiveWatermark(long watermark) {
+        lastEmittedWatermark = watermark;
         checkWatermarkAlignment();
     }
 
+    @Override
+    public void updateCurrentSplitWatermark(String splitId, long watermark) {
+        splitCurrentWatermarks.put(splitId, watermark);
+        if (currentMaxDesiredWatermark < watermark && !currentlyPausedSplits.contains(splitId)) {
+            pauseOrResumeSplits(Collections.singletonList(splitId), Collections.emptyList());
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java]: I agree, this is inconsistent in the case of a single split. I would add another check here for the number of splits. Calling checkSplitWatermarkAlignment is slightly more expensive due to it checking all splits.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:609]: I'll keep this as-is for now.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:627]: Yes, I had the wrong mental model when I wrote this. The source pausing is actually independent of the split alignment, even though both mechanism use the latest max allowed watermark.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 15]
```diff
@@ -562,11 +580,62 @@ private void updateMaxDesiredWatermark(WatermarkAlignmentEvent event) {
...
-        lastEmittedWatermark = emittedWatermark;
+    @Override
+    public void updateCurrentEffectiveWatermark(long watermark) {
+        lastEmittedWatermark = watermark;
         checkWatermarkAlignment();
     }
 
+    @Override
+    public void updateCurrentSplitWatermark(String splitId, long watermark) {
+        splitCurrentWatermarks.put(splitId, watermark);
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:591]: Actually, that's ok. It simply means that the split watermark of a split may not be registered if its watermark is always lagging behind another splits watermark. Either this split will overtake the other split and become registered here, or the other split will only ever get paused.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherPauseResumeSplitReaderTest.java (hunk 1 above)
Comment (mxm) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherPauseResumeSplitReaderTest.java:244]: This is not required. The signal simply allows for a clean shutdown of the source.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderBase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderBase.java]: Have you found an answer for that @mxm ?

File: flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReader.java [hunk 16]
```diff
@@ -161,6 +161,12 @@ default void notifyCheckpointComplete(long checkpointId) throws Exception {}
     default void pauseOrResumeSplits(
             Collection<String> splitsToPause, Collection<String> splitsToResume) {
         throw new UnsupportedOperationException(
-                "This source reader does not support pause or resume splits.");
+                "This split reader does not support pausing or resuming splits which can lead to unaligned splits.\n"
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReader.java:164]: nit: `This source reader`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 17]
```diff
@@ -589,10 +589,6 @@ public void updateCurrentEffectiveWatermark(long watermark) {
     @Override
     public void updateCurrentSplitWatermark(String splitId, long watermark) {
         splitCurrentWatermarks.put(splitId, watermark);
-        if (currentMaxDesiredWatermark < watermark && !currentlyPausedSplits.contains(splitId)) {
-            pauseOrResumeSplits(Collections.singletonList(splitId), Collections.emptyList());
-            currentlyPausedSplits.add(splitId);
-        }
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java]: Why are we removing this? If suddenly split wakes up and starts sending huge amount of records, lack of this code can cause problems.

edit: oh, you have reverted this change later, haven't you?

**Review by pnowojski — COMMENTED** [review]
Thanks for the update @mxm , I've left a couple of minor comments.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 17 above)
Comment (mxm) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java]: Yes, I didn't remove this but I fixed a bug which @mas-chen pointed out which would pause the split when there is just a single split.

**Comment by mxm** [comment]
> The next question would be how to squash the commits? If you want to preserve authorship of your changes @mxm, without overwriting authors of the base commits, we would need either a single fixup commit from you, all one your fixup per original commit.

I'd prefer this to be a single suqashed commit with the use of Co-Authored-By. IMHO there is no benefit in merging the individual commits because they are very much tight to each other. This keeps the history clean while retaining authorship. What do you think?

**Comment by pnowojski** [comment]
> I'd prefer this to be a single suqashed commit with the use of Co-Authored-By. IMHO there is no benefit in merging the individual commits because they are very much tight to each other. This keeps the history clean while retaining authorship. What do you think?

I wouldn't squash all commits into one. The original commits authored by Sebastian, Dawid and Arvid I think are nicely separated. But sure, squashing your fixups to those commits and co-authoring them is also perfectly fine.

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java [hunk 18]
```diff
@@ -95,10 +97,12 @@ public boolean updateCombinedWatermark() {
     static class PartialWatermark {
         private long watermark = Long.MIN_VALUE;
         private boolean idle = false;
+
+        @Nullable
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java:101]: If this is only used for the sake of tests, I would actually prefer to leave this field as non-null with no-op lambda functions.

Comment (mxm) [reply flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java]: Agreed.

**Comment by mxm** [comment]
@flinkbot run azure

**Review by pnowojski — APPROVED** [review]
We are often splitting larger features into smaller commits for the sake of easier reviewing. As long as commits are independently working, and coherent (one need to find a sensible division) that's fine.

https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#separate-refactoring-cleanup-and-independent-changes

It gets more complicated in those rare cases where there are various authors, especially if someone takes over an orphaned PR. 🤷 

LGTM assuming green azure.

**Comment by smattheis** [comment]
@mxm Thanks a lot also from my side for you contributions. However, @pnowojski this PR was not orphaned! (I wrote to you that I'm on vacations.)

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderBase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderBase.java]: Yes, Pulsar splits are always started in a separate fetcher thread. We still want to be able to pause these split threads. So the code makes sense.

**Comment by mxm** [comment]
Thank you @pnowojski for your thoughtful comments and guidance. And of course thanks to the original authors @AHeise @dawidwys @smattheis.

Whoops @smattheis I had assumed you wouldn't continue on this PR but I think no harm has been done. Let me know if anything is open on your side. It still need to revise the code with respect to a test failure.

**Comment by pnowojski** [comment]
> @mxm Thanks a lot also from my side for you contributions. However, @pnowojski this PR was not orphaned! (I wrote to you that I'm on vacations.)

Yes, sorry for the mental shortcut. There was a delay in the communication between first my vacations and then yours. Around the time @mxm was asking how to continue with this PR, you @smattheis were not responding, and I got a message from you that you can work on it only after @mxm has already volunteered to take it over. After that I didn't want to create more confusion. But sorry for my inaccurate statement that it was "orphaned" :)

**Comment by smattheis** [comment]
@mxm @pnowojski All good. Thanks for driving it.

**Comment by mxm** [comment]
(Rebased)

**Comment by mxm** [comment]
@flinkbot run azure

**Comment by pnowojski** [comment]
There are still some related test failures:
> Documented option pipeline.watermark-alignment.allow-unaligned-source-splits does not exist.

and
> Sep 09 15:10:03 [ERROR] Errors: 
> Sep 09 15:10:03 [ERROR]   PulsarUnorderedSourceReaderTest.supportsPausingOrResumingSplits(PulsarSourceReaderBase, Boundedness, String)[1] » Timeout
Sep 09 15:10:03 [ERROR]   PulsarUnorderedSourceReaderTest.supportsPausingOrResumingSplits(PulsarSourceReaderBase, Boundedness, String)[2] » Timeout

**Comment by mxm** [comment]
Thanks @pnowojski. Will fix those. 

On another note, do you think it makes sense to add a configuration option to disable using split alignment entirely? I'm a bit worried this feature might cause issues with the already integrated Kafka/Pulsar sources that we are not foreseeing right now. The added configuration option only allows to disable split alignments for non-compatible connectors. We could add another option to disable it entirely to retain the old behavior (disabled by default).

**Comment by mxm** [comment]
> There are still some related test failures:
> 
> > Documented option pipeline.watermark-alignment.allow-unaligned-source-splits does not exist.

I had to remove the `@Deprecated` flag from the configuration option. Otherwise, it will be undocumented.

**Comment by mxm** [comment]
@flinkbot run azure

**Comment by pnowojski** [comment]
> On another note, do you think it makes sense to add a configuration option to disable using split alignment entirely? I'm a bit worried this feature might cause issues with the already integrated Kafka/Pulsar sources that we are not foreseeing right now. The added configuration option only allows to disable split alignments for non-compatible connectors. We could add another option to disable it entirely to retain the old behavior (disabled by default).

Does it make sense to use and enable watermark alignment via `DataStream API` to only disable it via config value? Keep in mind that user has to manually enable via `WatermarkStrategy#withWatermarkAlignment` call [1].

[1] https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/event-time/generating_watermarks/#watermark-alignment-_beta_

**Comment by mxm** [comment]
Fair point, users may modify the job to disable alignment. It doesn't make sense to override the watermark alignment settings from the config.

**Comment by mxm** [comment]
Fair point, users may modify the job to disable alignment. It doesn't make sense to override the watermark alignment settings from the config. But the same argument applies to the other configuration option to disable split alignment for not-ready sources. I think it would make sense to provide a transitory option to disable split alignment altogether until this feature has proven to be stable.

**Comment by mxm** [comment]
@flinkbot run azure

**Comment by mxm** [comment]
Tests are passing now.

**Review by mxm — APPROVED** [review]

**Comment by mxm** [comment]
Thank you everyone!

**Comment by dawidwys** [comment]
Thank you for taking over! @mxm

**Comment by tweise** [comment]
Exciting to see this make the finish line! This was a long awaited feature (we implemented alignment for the Kinesis source ~4 years ago). Let's see how it holds up in production use cases!

Thanks @mxm @pnowojski @smattheis  @dawidwys, @AHeise for your contributions!

**Comment by pnowojski** [comment]
Yes, it took longer than it should have. Good that it's finally been merged 🎉
